src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,181 @@ namespace mlpack {
 /**
  * Given a dataset, create another dataset via bootstrap sampling, with labels.
  */
-template<bool UseWeights,
-         typename MatType,
-         typename LabelsType,
-         typename WeightsType>
-void Bootstrap(const MatType& dataset,
-               const LabelsType& labels,
-               const WeightsType& weights,
-               MatType& bootstrapDataset,
-               LabelsType& bootstrapLabels,
-               WeightsType& bootstrapWeights)
+class DefaultBootstrap
 {
-  bootstrapDataset.set_size(dataset.n_rows, dataset.n_cols);
-  bootstrapLabels.set_size(labels.n_elem);
-  if (UseWeights)
-    bootstrapWeights.set_size(weights.n_elem);
+ public:
+  template<bool UseWeights,
+           typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    bootstrapDataset.set_size(dataset.n_rows, dataset.n_cols);
+    bootstrapLabels.set_size(labels.n_elem);
+    if (UseWeights)
+      bootstrapWeights.set_size(weights.n_elem);
 
-  // Random sampling with replacement.
-  arma::uvec indices = randi<arma::uvec>(dataset.n_cols,
+    // Random sampling with replacement.
+    arma::uvec indices = randi<arma::uvec>(dataset.n_cols,
       arma::distr_param(0, dataset.n_cols - 1));
-  bootstrapDataset = dataset.cols(indices);
-  bootstrapLabels = labels.cols(indices);
-  if (UseWeights)
-    bootstrapWeights = weights.cols(indices);
-}
+    bootstrapDataset = dataset.cols(indices);
+    bootstrapLabels = labels.cols(indices);
+    if (UseWeights)
+      bootstrapWeights = weights.cols(indices);
+  }
+};
+
+/**
+ * No-op bootstrap strategy that just copies the input to the output.
+ */
+class IdentityBootstrap
+{
+ public:
+  template<bool UseWeights,
+           typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    MakeAlias(bootstrapDataset, dataset, dataset.n_rows, dataset.n_cols);
+    MakeAlias(bootstrapLabels, labels, labels.n_elem);
+    if (UseWeights)
+      MakeAlias(bootstrapWeights, weights, weights.n_elem);
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ */
+template<typename IndMatType = arma::mat>
+class SequentialBootstrap
```

Comment:
Thanks again for the description of how the sequential bootstrap works.  In this view I don't seem to see that conversation (I guess it must be associated with an old diff), so I'll just start a new conversation...

Under the assumption that intervals are contiguous (e.g. they can be parameterized by a start sample and an end sample), it's pretty memory-inefficient (and to me counterintuitive) to represent them as a big, sparse indicator matrix.  That matrix will have size `n` x `m`, where `m` is the number of intervals.  Thus anything we do with that matrix gives us an `O(nm)` runtime.

To me, the following API would make a lot more sense:

```
// `intervals` should be a `2 x m` matrix, where each column has the start sample and end
// sample of an interval.
SequentialBootstrap(const arma::umat& intervals = arma::umat());

// not sure I got the default parameter above correct, maybe it needs two constructors,
// but if the user doesn't specify intervals, then we can assume there is only one interval
// [0, n - 1], in which case this is just the regular bootstrap
```

This representation has size `2m`, doesn't have any zero elements, and using it, it is possible to compute the uniqueness of each sample in a two-pass algorithm in `O(m log m) + O(n)` time.  Basically you would sort the intervals by start sample, then iterate over them, keeping track of how many intervals are "active" at a given sample.  I'm handwaving but if you want me to clarify what I mean just let me know.

Especially for random forests, I expect there to often be quite a lot of data, and so being able to efficiently bootstrap is important.

If my assumption about contiguous intervals is wrong, we could:
 * also provide an indicator matrix interface (I don't like this as much, it's still inefficient)
 * I believe the computation is unchanged if you treat a non-contiguous interval as multiple separate contiguous intervals, so we can just tell users to do that
 * or we can accept a structure like a `std::vector<arma::umat>` where each `umat` in the `std::vector` is a 2-row matrix that has all of the intervals for a given sequence

I know this amounts to a big change to what you've got, but I think it's really important to provide a generic and efficient tool that's easy to understand.  I get the impression the indicator matrix might be a common thing in financial machine learning, but those aren't the only users of mlpack and I have not encountered that in any machine learning I have done.  If we can give non-financial machine learning users an easy way to onboard this technique, it will get more adoption too.

Reply:
@rcurtin , sorry I was also busy the last weeks. I really like this proposal of yours! I'll implement it tomorrow together with the other documentation remarks that you made. Then you can review again. Really appreciate your feedback!!